src/interpreter/s2s_forward_mode_ad.jl

```diff
@@ -137,27 +118,39 @@ function modify_fwd_ad_stmts!(
 end
 
 function modify_fwd_ad_stmts!(
-    dual_ir::IRCode,
+    dual_ir::CC.IncrementalCompact,
     primal_ir::IRCode,
     ::MooncakeInterpreter,
-    stmt::Union{GotoNode,Core.GotoIfNot},
+    stmt::GotoNode,
     i::Integer;
     kwargs...,
 )
     return nothing
 end
 
 function modify_fwd_ad_stmts!(
-    dual_ir::IRCode,
+    dual_ir::CC.IncrementalCompact,
     primal_ir::IRCode,
     ::MooncakeInterpreter,
-    stmt::GotoIfNot,
+    stmt::Core.GotoIfNot,
     i::Integer;
     kwargs...,
 )
-    return Mooncake.replace_call!(
-        dual_ir, CC.SSAValue(i), Core.GotoIfNot(CC.SSAValue(i - 1), stmt.dest)
+    # replace GotoIfNot with the call to primal
+    Mooncake.replace_call!(dual_ir, CC.SSAValue(i), Expr(:call, _primal, stmt.cond))
+    # reinsert the GotoIfNot right after the call to primal
+    # (incremental insertion cannot be done before "where we are")
+    new_gotoifnot_inst = CC.NewInstruction(
+        Core.GotoIfNot(CC.SSAValue(i), stmt.dest),  #
+        Any,
+        CC.NoCallInfo(),
+        Int32(1),  # meaningless
```

Comment:
You likely want `nothing` since that will try to figure out the right line number instead of just `1`.

Reply:
We tried with nothing but the incremental compact insertion threw an error